ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs

```diff
@@ -73,20 +81,121 @@ import qualified Shelley.Spec.Ledger.UTxO as SL
+
+    cpuSlot :: CC.Update.CandidateProtocolUpdate -> SlotNo
+    cpuSlot = Byron.fromByronSlotNo . CC.Update.cpuSlot
+
+    -- This follows the same structure as the computation in the A/B test. Let
+    -- @s@ be the slot of the update proposal. Note that the very first slot in
+    -- which the transition /could/ occur is @s + 1@; adding the required
+    -- stability, the first slot in which the transition could occur is @s + 4k
+    -- + 1@. This means that the last slot which /must/ be in /this/ era is @s +
+    -- 4k@. Hence the last /epoch/ that must be in this era is @epoch (s + 4k)@,
+    -- and the first epoch of the /next/ era is @succ (epoch (s + 4k)@.
+    cpuEpoch :: CC.Update.CandidateProtocolUpdate -> EpochNo
+    cpuEpoch upd = succ (slotToEpoch $ History.addSlots (4 * k) (cpuSlot upd))
+
+    -- Slot conversion (valid for slots in this era only)
+    slotToEpoch :: SlotNo -> EpochNo
+    slotToEpoch s =
+        History.addEpochs
+          (History.countSlots s (boundSlot eraStart) `div` unEpochSize eraEpochSize)
+          (boundEpoch eraStart)
+
+    isStable :: CC.Update.CandidateProtocolUpdate -> Bool
+    isStable upd = confirmationDepth >= 2 * k
+      where
+        confirmedInSlot :: SlotNo
+        confirmedInSlot = cpuSlot upd
+
+        -- The impossible cases are impossible because the ledger contains
+        -- at least the block containing the update proposal, and hence cannot
+        -- be empty or have a tip /before/ the slot number of that block.
+        confirmationDepth :: Word64
+        confirmationDepth =
+            case ledgerTipSlot st of
+              Origin -> error "byronTransition: impossible"
+              At s   -> if s < confirmedInSlot
+                          then error "byronTransition: impossible"
+                          else History.countSlots s confirmedInSlot
+
+    -- 'tryBumpVersion' assumes head of the list is the newest, so we do too
+    latest :: [CC.Update.CandidateProtocolUpdate]
+           -> Maybe CC.Update.CandidateProtocolUpdate
+    latest (newest:_) = Just newest
+    latest []         = Nothing
+
 {-------------------------------------------------------------------------------
   SingleEraBlock Byron
 -------------------------------------------------------------------------------}
 
 instance SingleEraBlock ByronBlock where
-  -- | TODO Currently we hard-code the transition in the partial ledger config
-  -- for testing purposes.
-  --
-  -- This will be replaced with a proper implementation that looks for the
-  -- right transaction in the ledger. See:
-  -- <https://github.com/input-output-hk/ouroboros-network/issues/1786>
-  singleEraTransition pcfg _ledgerState =
+  singleEraTransition pcfg eraParams eraStart ledgerState =
       case transitionEpoch pcfg of
-        NoHardCodedTransition       -> Nothing
+        NoHardCodedTransition -> byronTransition pcfg eraParams eraStart ledgerState
         HardCodedTransitionAt epoch -> Just epoch
```

Comment:
Your friendship with @mrBliss is at risk with this indentation.

Reply:
Alignment vs. a too long line. I think I would indeed have kept the arrows aligned but moved the RHSs to a newline :grin:

Reply:
... but then you'd have code in between two aligned lines and you lose the benefit of alignment. Life is _sooo_ hard :(